The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3490006353/jobs/5840863134) [(plain)](https://github.com/rust-lang/rust/commit/e871963a5a8ca594aa58a0f2aa9f76c979fd89e9/checks/9555147319/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

Some tests failed in compiletest suite=codegen mode=codegen host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
---- [codegen] src/test/codegen/enum-match.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-13/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/enum-match/enum-match.ll" "/checkout/src/test/codegen/enum-match.rs" "--allow-unused-prefixes" "--check-prefixes" "CHECK,NONMSVC"
stdout: none
--- stderr -------------------------------
/checkout/src/test/codegen/enum-match.rs:37:16: error: CHECK-NEXT: expected string not found in input
// CHECK-NEXT: %1 = tail call i8 @llvm.usub.sat.i8(i8 %0, i8 1)
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/enum-match/enum-match.ll:17:7: note: scanning from here
      ^
      ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/enum-match/enum-match.ll:18:2: note: possible intended match here
 %1 = call i8 @llvm.usub.sat.i8(i8 %0, i8 1)

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/enum-match/enum-match.ll
Check file: /checkout/src/test/codegen/enum-match.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
           .
           .
           .
           .
          12:  ret i8 %.0 
          13: } 
          14:  
          15: ; Function Attrs: mustprogress nofree nosync nounwind nonlazybind readnone uwtable willreturn 
          16: define i8 @match1(i8 noundef %0) unnamed_addr #1 { 
          17: start: 
next:37'0           X error: no match found
          18:  %1 = call i8 @llvm.usub.sat.i8(i8 %0, i8 1) 
next:37'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
next:37'1      ?                                            possible intended match
          19:  switch i8 %1, label %bb2 [ 
next:37'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          20:  i8 0, label %bb3 
next:37'0     ~~~~~~~~~~~~~~~~~~
          21:  i8 1, label %bb5 
next:37'0     ~~~~~~~~~~~~~~~~~~
          22:  i8 2, label %bb1 
next:37'0     ~~~~~~~~~~~~~~~~~~
          23:  ] 
next:37'0     ~~~
           .
           .
>>>>>>
------------------------------------------
```

</details>

--- a/compiler/rustc_codegen_ssa/src/mir/place.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/place.rs
@@ -309,14 +309,14 @@ impl<'a, 'tcx, V: CodegenObject> PlaceRef<'tcx, V> {
                 // In the algorithm above, we can change
                 // cast(relative_tag) + niche_variants.start()
                 // into
-                // cast(tag) + (niche_variants.start() - niche_start)
+                // cast(tag + (niche_variants.start() - niche_start))
                 // if either the casted type is no larger than the original
                 // type, or if the niche values are contiguous (in either the
                 // signed or unsigned sense).
-                let can_incr_after_cast = cast_smaller || niches_ule || niches_sle;
+                let can_incr = cast_smaller || niches_ule || niches_sle;
 
                 let data_for_boundary_niche = || -> Option<(IntPredicate, u128)> {
-                    if !can_incr_after_cast {
+                    if !can_incr {
                         None
                     } else if niche_start == low_unsigned {
                         Some((IntPredicate::IntULE, niche_end))
@@ -353,24 +353,33 @@ impl<'a, 'tcx, V: CodegenObject> PlaceRef<'tcx, V> {
                     // The algorithm is now this:
                     // is_niche = tag <= niche_end
                     // discr = if is_niche {
-                    //     cast(tag) + (niche_variants.start() - niche_start)
+                    //     cast(tag + (niche_variants.start() - niche_start))
                     // } else {
                     //     untagged_variant
                     // }
                     // (the first line may instead be tag >= niche_start,
                     // and may be a signed or unsigned comparison)
+                    // The arithmetic must be done before the cast, so we can
+                    // have the correct wrapping behavior. See issue #104519 for
+                    // the consequences of getting this wrong.
                     let is_niche =
                         bx.icmp(predicate, tag, bx.cx().const_uint_big(tag_llty, constant));
-                    let cast_tag = if cast_smaller {
-                        bx.intcast(tag, cast_to, false)
-                    } else if niches_ule {
-                        bx.zext(tag, cast_to)
+                    let delta = (niche_variants.start().as_u32() as u128).wrapping_sub(niche_start);
+                    let incr_tag = if delta == 0 {
+                        tag
                     } else {
-                        bx.sext(tag, cast_to)
+                        bx.add(tag, bx.cx().const_uint_big(tag_llty, delta))
                     };
 
-                    let delta = (niche_variants.start().as_u32() as u128).wrapping_sub(niche_start);
-                    (is_niche, cast_tag, delta)
+                    let cast_tag = if cast_smaller {
+                        bx.intcast(incr_tag, cast_to, false)
+                    } else if niches_ule {
+                        bx.zext(incr_tag, cast_to)
+                    } else {
+                        bx.sext(incr_tag, cast_to)
+                    };
+
+                    (is_niche, cast_tag, 0)
                 } else {
                     // The special cases don't apply, so we'll have to go with
                     // the general algorithm.

--- a/src/test/codegen/enum-match.rs
+++ b/src/test/codegen/enum-match.rs
@@ -34,11 +34,8 @@ pub enum Enum1 {
 
 // CHECK: define i8 @match1{{.*}}
 // CHECK-NEXT: start:
-// CHECK-NEXT: %1 = icmp ugt i8 %0, 1
-// CHECK-NEXT: %2 = zext i8 %0 to i64
-// CHECK-NEXT: %3 = add nsw i64 %2, -1
-// CHECK-NEXT: %_2 = select i1 %1, i64 %3, i64 0
-// CHECK-NEXT: switch i64 %_2, label {{.*}} [
+// CHECK-NEXT: %1 = {{.*}}call i8 @llvm.usub.sat.i8(i8 %0, i8 1)
+// CHECK-NEXT: switch i8 %1, label {{.*}} [
 #[no_mangle]
 pub fn match1(e: Enum1) -> u8 {
     use Enum1::*;

--- a/src/test/ui/enum-discriminant/issue-104519.rs
+++ b/src/test/ui/enum-discriminant/issue-104519.rs
@@ -1,0 +1,36 @@
+// run-pass
+#![allow(dead_code)]
+
+enum OpenResult {
+    Ok(()),
+    Err(()),
+    TransportErr(TransportErr),
+}
+
+#[repr(i32)]
+enum TransportErr {
+    UnknownMethod = -2,
+}
+
+#[inline(never)]
+fn some_match(result: OpenResult) -> u8 {
+    match result {
+        OpenResult::Ok(()) => 0,
+        _ => 1,
+    }
+}
+
+fn main() {
+    let result = OpenResult::Ok(());
+    assert_eq!(some_match(result), 0);
+
+    let result = OpenResult::Ok(());
+    match result {
+        OpenResult::Ok(()) => (),
+        _ => unreachable!("message a"),
+    }
+    match result {
+        OpenResult::Ok(()) => (),
+        _ => unreachable!("message b"),
+    }
+}

